src/validation.cpp

```diff
@@ -6410,6 +6415,20 @@ std::optional<int> ChainstateManager::GetSnapshotBaseHeight() const
     return base ? std::make_optional(base->nHeight) : std::nullopt;
 }
 
+void ChainstateManager::RecalculateBestHeader()
+{
+    AssertLockHeld(cs_main);
+    // If, due to invalidation / reconsideration of blocks, the previous
+    // best header is no longer valid or guaranteed to be the most-work
+    // header in our block-index not known to be invalid, recalculate it.
+    m_best_header = ActiveChain().Tip();
+    for (auto& entry : m_blockman.m_block_index) {
+        if (!(entry.second.nStatus & BLOCK_FAILED_MASK) && m_best_header->nChainWork < entry.second.nChainWork) {
+            m_best_header = &entry.second;
+        }
+    }
```

Comment:
I've been looking the logic over again and think it might be a good idea to call `NotifyHeaderTip` here too. It basically tracks along `m_best_header`, but with the current code will only notify when processing the next block or header.

Reply:
Makes sense. I've added a call to `NotifyHeaderTip`.

Reply:
actually, that didn't work because `NotifyHeaderTip` can't be called with `cs_main` is held. Reverting this, I'll think about this a bit more.

Reply:
Ah right, I don't think this a blocker for this PR anyway, just something that came to my mind.

Reply:
I thought about it a bit and don't see a great non-invasive solution to this: We could add a call to `NotifyHeaderTip` in the calling functions after  `cs_main` is released, which would be in `ActivateBestChain()` (plus in the `invalidateblock` / `reconsiderblock` rpcs which seem less important), and I'm unsure if we want to refactor this function to always call `NotifyBestHeader()` just for the special case where a connected block turns out to be invalid.
In any case, since `m_best_header` is already being reset in `InvalidChainFound()` in master (just not always correctly), so this would not be a regression introduced by this PR, so I think it's ok not to address it here for now.

Reply:
各方面看过嗯结果OK